tools/buildgen/extract_metadata_from_bazel_xml.py

```diff
@@ -247,119 +244,197 @@ def _external_dep_name_from_bazel_dependency(bazel_dep):
+
+    The final build metadata are:
+    * _TRANSITIVE_DEPS: all the transitive dependencies including intermediate
+                        targets;
+    * _COLLAPSED_DEPS:  dependencies that fits our requirement above, and it
+                        will remove duplicated items and produce the shortest
+                        possible dependency list in alphabetical order;
+    * _COLLAPSED_SRCS:  the merged source files;
+    * _COLLAPSED_PUBLIC_HEADERS: the merged public headers;
+    * _COLLAPSED_HEADERS: the merged non-public headers.
+    * _EXCLUDE_DEPS: the intermediate dependency labels.
+
+    For the collapsed_deps, the algorithm improved cases like:
+
+    The result in the past:
+        end2end_tests -> [grpc_test_util, grpc, gpr, address_sorting, upb]
+        grpc_test_util -> [grpc, gpr, address_sorting, upb, ...]
+        grpc -> [gpr, address_sorting, upb, ...]
+    
+    The result of the algorithm:
+        end2end_tests -> [grpc_test_util]
+        grpc_test_util -> [grpc]
+        grpc -> [gpr, address_sorting, upb, ...]
+    """
+    bazel_rule = bazel_rules[rule_name]
+    direct_deps = _extract_deps(bazel_rule, bazel_rules)
+    transitive_deps = []
+    # We care about the order of dependencies list, so it will be a list.
+    collapsed_deps = []
+    # We don't care about the order of srcs and headers, they will be sorted.
+    collapsed_srcs = set(_extract_sources(bazel_rule))
+    collapsed_public_headers = set(_extract_public_headers(bazel_rule))
+    collapsed_headers = set(_extract_nonpublic_headers(bazel_rule))
+    exclude_deps = set()
+
+    for dep in direct_deps:
+        external_dep_name_maybe = _external_dep_name_from_bazel_dependency(dep)
+
+        if dep in bazel_rules:
+            # Suppress expanding absl libs; expanding absl deps are technically
+            # correct but creates too much noise
+            if external_dep_name_maybe is None or not external_dep_name_maybe.startswith(
+                    'absl'):
+                if "_PROCESSING_DONE" not in bazel_rules[dep]:
+                    # This item is not processed before, compute now
+                    _compute_transitive_metadata(dep, bazel_rules,
+                                                 bazel_label_to_dep_name)
+                _deduplicate_append(
+                    transitive_deps,
+                    bazel_rules[dep].get('_TRANSITIVE_DEPS', []))
+                _deduplicate_append(collapsed_deps,
+                                    bazel_rules[dep].get('_COLLAPSED_DEPS', []))
+                exclude_deps.update(bazel_rules[dep].get('_EXCLUDE_DEPS', []))
+
+        # This dep is a public target, add it as a dependency
+        if dep in bazel_label_to_dep_name:
+            _deduplicate_append(transitive_deps, [bazel_label_to_dep_name[dep]])
+            _deduplicate_append(collapsed_deps, [bazel_label_to_dep_name[dep]])
+            exclude_deps.update(bazel_rules[dep]['_TRANSITIVE_DEPS'])
```

Comment:
I think if you added a comment here `# add all the transitive deps of our every public dep to exclude list since we want to avoid building sources that are already built our dependencies` it would make the intent clearer?

Reply:
Good idea! Added.